Updated getHoldDuration

fedorovvldmr opened the pull request.

Alex7Kom commented:
Can you please elaborate what do you want to achieve?

fedorovvldmr commented:
english: The call "eval" works faster (at least on my server with node.js 8.9.4) and with "use strict" it has its own scope
russian: Вызов "eval" работает быстрее (по крайней мере на моем сервере с node.js 8.9.4) и с "use strict" имеет свою область видимости

Alex7Kom commented:
It does have its own scope but it also has access to everything the surrounding code can access so Valve code can break the module or your app by accident. Even with an indirect eval call (`(0, eval)('code')`) it has access to the global object.

While `eval` is indeed faster, I think we can reuse the context for vm and gain some speed.

```
vm#runInNewContext x 1,318 ops/sec ±6.20% (70 runs sampled)
eval x 2,151,221 ops/sec ±6.52% (75 runs sampled)
vm#runInContext x 27,996 ops/sec ±6.04% (76 runs sampled)
Fastest is eval
```

```js
var Benchmark = require('benchmark');

var suite = new Benchmark.Suite;

var vm = require('vm');
var code = `
var g_daysMyEscrow = 0;
var g_daysTheirEscrow = 1;
var g_daysBothEscrow = 1;

var data = {
  my: g_daysMyEscrow,
  their: g_daysTheirEscrow
};
`;

var sandbox = vm.createContext();

// add tests
suite.add('vm#runInNewContext', function() {
  vm.runInNewContext(code);
})
.add('eval', function() {
  "use strict";

  eval(code);
})
.add('vm#runInContext', function() {
  vm.runInContext(code, sandbox);
})
// add listeners
.on('cycle', function(event) {
  console.log(String(event.target));
})
.on('complete', function() {
  console.log('Fastest is ' + this.filter('fastest').map('name'));
})
// run async
.run({ 'async': true });

```

Tell me if this'll work for you.

P.S. Sorry, I'd like issues for my OS projects to be in English so anyone can contribute. Feel free to use Google Translate, though!

fedorovvldmr commented:
I changed the function because in my application the "getHolDuration" worked about 3-4 seconds. I did not like it.
I forgot to replace "var" with "let". This completely solves the problem of accessing variables to the outer scope. I'm not a very good and experienced programmer. Is this a good solution?

For the interest i changed the code. Added code:
`.add('RegExp', function() {
    code.match(/var\s+g_daysMyEscrow\s+=\s+([0-9]+);/)
})`

Result:
> vm#runInNewContext x 1,152 ops/sec 10.66% (61 runs sampled)
eval x 1,473,653 ops/sec 3.71% (60 runs sampled)
vm#runInContext x 20,521 ops/sec 8.47% (64 runs sampled)
RegExp x 3,838,068 ops/sec 7.14% (60 runs sampled)
Fastest is RegExp

Alex7Kom commented:
> I changed the function because in my application the "getHolDuration" worked about 3-4 seconds. I did not like it.

Решил ли проблему `eval`? Сколько работает эта функция с `eval`? `vm. runInNewContext` не должен занимать больше 10-20 мс в среднем. Может проблема в другом?

> I forgot to replace "var" with "let". This completely solves the problem of accessing variables to the outer scope. I'm not a very good and experienced programmer. Is this a good solution?

`eval` даже с `"use strict"` имеет доступ ко всему, к чему имеет доступ окружающий его код.

Единственное отличие `eval` с `"use strict"` от обычного `eval` — наличие собственной области видимости, поэтому переменные, объявленные внутри `eval`, не видны снаружи:

```js
(function testEval () {
  "use strict";

  var x = 0;
  var y = 1;
  eval('var y = 5; x = 10; var z = 15;');

  console.log(x); // 10
  console.log(y); // 1
  console.log(z); // ReferenceError: z is not defined
})();
```

Однако значение `x` снаружи успешно поменялось.

Это проблема, так как код Valve таким образом сможет менять переменные внутри нашего приложения и случайно сломать его.

Вместо этого используется модуль [`vm`](https://nodejs.org/api/vm.html), и код бежит в изолированном контексте от нашего приложения. Создание контекста занимает время, поэтому `vm#runInNewContext` медленный. Мы можем создавать контекст 1 раз при первом обращении или запуске приложений, и дальше использовать его повторно с помощью `vm#runInContext`, что в 20 раз быстрее.

```
vm#runInNewContext x 1,318 ops/sec ±6.20% (70 runs sampled)
eval x 2,151,221 ops/sec ±6.52% (75 runs sampled)
vm#runInContext x 27,996 ops/sec ±6.04% (76 runs sampled)
Fastest is eval
```

> For the interest i changed the code. Added code:
.add('RegExp', function() { code.match(/var\s+g_daysMyEscrow\s+=\s+([0-9]+);/) })

Я хотел бы по возможности избежать использования RegExp, так как такой код хрупкий.

fedorovvldmr commented:
Вариант с eval выполнялся максимум за 0,4 секунды в худших случаях. По моим измерениям основная задержка была в vm.runInNewContext(code, sandbox). Нагрузка на сервер была небольшая. Сервер OpenVZ SSD VPS на Debian 8 c 2 ярами и 4 Гб оперативной памяти.
